appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/OpenApiService.java

```diff
@@ -251,4 +239,70 @@ private static String getContextRoot(ApplicationInfo appInfo) {
                 .collect(toSet());
     }
 
+    private class OpenApiMapping {
+
+        private final ApplicationInfo appInfo;
+        private final OpenApiConfiguration appConfig;
+        private final CompletableFuture<OpenAPI> document;
+
+        private OpenApiMapping(ApplicationInfo appInfo) {
+            this.appInfo = appInfo;
+            this.appConfig = new OpenApiConfiguration(appInfo.getAppClassLoader());
+            this.document = new CompletableFuture<>();
+        }
+
+        private ApplicationInfo getAppInfo() {
+            return appInfo;
+        }
+
+        private OpenApiConfiguration getAppConfig() {
+            return appConfig;
+        }
+
+        private synchronized OpenAPI getDocument() throws InterruptedException, ExecutionException, TimeoutException {
+            if (!document.isDone()) {
+                executor.submit(new OpenApiBuilder(this, document));
+            }
+            return document.get(10, TimeUnit.SECONDS);
+        }
+
+    }
+
+    private class OpenApiBuilder implements Runnable {
+
+        private final OpenApiMapping mapping;
+        private final CompletableFuture<OpenAPI> future;
+
+        private OpenApiBuilder(OpenApiMapping mapping, CompletableFuture<OpenAPI> future) {
+            this.mapping = mapping;
+            this.future = future;
+        }
+
+        @Override
+        public void run() {
+            if (!future.isDone()) {
```

Comment:
This check is not thread safe - if the executor has more threads, the run() method can be executed concurrently. Before the first one is finished, the future isn't done so the second one will also try to do the job and they will both run in parallel.

The call to future.isDone() isn't atomic and the future might be being completed in a parallel thread while not yet done.